### PR TITLE
feat: NeMo Gym refresh 20260113

### PR DESCRIPTION
**Pull request by bxyu-nvidia**
# What does this PR do ?

1. Refactor Gym workspace from stub module to true Gym repo
   1. 3rdparty/Gym-workspace/is_nemo_gym_installed.py
   2. 3rdparty/Gym-workspace/pyproject.toml
   3. 3rdparty/Gym-workspace/setup.py
   4. pyproject.toml
   5. uv.lock
2. Clean remaining "Penguin" references
   1. docs/design-docs/dependency-management.md
3. Clean and update Gym config and launch files
   1. examples/nemo_gym/grpo_dapo17k_bytedtsinghua_qwen3_4binstruct_nf.yaml
   2. examples/nemo_gym/grpo_qwen3_30ba3b_instruct.yaml
   3. examples/nemo_gym/grpo_workplace_assistant_nemotron_nano_v2_9b.yaml
   4. examples/nemo_gym/launch_nemo_gym_multinode_training.sh
   5. nemo_rl/environments/nemo_gym.py
4. O(n * log(n)) bin packing algorithm - significantly speeds up larger rollout batch size jobs
   1. nemo_rl/data/packing/algorithms.py
5. Fix off-by-one logging problem
   1. `step_finished` logic in nemo_rl/utils/logger.py and nemo_rl/algorithms/grpo.py
6. Upload raylet.out and raylet.err files to W&B
   1. raylet logic in nemo_rl/utils/logger.py
7. Support for logging responses to disk. Saves several mins for larger rollout batch size jobs
   1. `log_string_list_as_jsonl` logic in nemo_rl/utils/logger.py
8. CPU and Ray memory tracking util
   1. nemo_rl/utils/memory_tracker.py
   2. nemo_rl/algorithms/grpo.py
9. Opt-out NeMo Gym logging to W&B
   1. `should_log_nemo_gym_responses` logic in nemo_rl/algorithms/grpo.py
10. Opt-in skip reference logprobs calculation when KL penalty is 0
    1. `skip_reference_policy_logprobs_calculation` logic in nemo_rl/algorithms/grpo.py
    2. nemo_rl/algorithms/loss_functions.py
11. Support advantage calculation on GPU
    1. `calculate_advantages_on_gpu` in nemo_rl/algorithms/grpo.py
12. Reduce CPU usage in GRPO loop
    1. `del *` variable deletions in nemo_rl/algorithms/grpo.py
13. Log mixed reward
    1. `baseline_reward/pct_0` related logic in nemo_rl/algorithms/grpo.py

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

## Release Notes

* **New Features**
  * GRPO algorithm now supports GPU-based advantage calculations and optional response logging.
  * Added memory tracking instrumentation for training pipeline stages.
  * New JSONL logging capability for structured metrics output.
  * Enhanced multinode training script with SLURM node count configuration.
  * New Qwen3 30B GRPO configuration example.

* **Documentation**
  * Updated dependency management guide.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: examples/nemo_gym/grpo_workplace_assistant_nemotron_nano_v2_9b.yaml [hunk 1]
```diff
@@ -201,7 +201,7 @@ policy:
       kv_cache_dtype: "auto"
       expose_http_server: true
       skip_tokenizer_init: false
-      tool_parser_plugin: ???
+      # tool_parser_plugin: ???  # This is set to the path for Nemotron Nano v2
```
Comment (yfw) [inline examples/nemo_gym/grpo_workplace_assistant_nemotron_nano_v2_9b.yaml:204]: Is this supposed to be commented out?

File: nemo_rl/utils/logger.py [hunk 2]
```diff
@@ -199,6 +201,14 @@ class WandbLogger(LoggerInterface):
 
     def __init__(self, cfg: WandbConfig, log_dir: Optional[str] = None):
         self.run = wandb.init(**cfg, dir=log_dir)
+
+        if os.environ.get("RAY_BACKEND_LOG_LEVEL", "").lower() == "debug":
+            print(
+                "Uploading raylet.out and raylet.err files to W&B since environment variable RAY_BACKEND_LOG_LEVEL=debug"
+            )
+            wandb.save("/tmp/ray/session_latest/logs/raylet.out", policy="live")
```
Comment (yfw) [inline nemo_rl/utils/logger.py:209]: nit: is there any way to get this directory programmatically?

File: nemo_rl/utils/logger.py [hunk 3]
```diff
@@ -954,6 +972,24 @@ def log_batched_dict_as_jsonl(
 
         print(f"Logged data to {filepath}")
 
+    def log_string_list_as_jsonl(self, to_log: list[str], filename: str) -> None:
```
Comment (yfw) [inline nemo_rl/utils/logger.py:975]: Might have missed this, but where is this used?

File: examples/nemo_gym/grpo_workplace_assistant_nemotron_nano_v2_9b.yaml (hunk 1 above)
Comment (bxyu-nvidia) [reply examples/nemo_gym/grpo_workplace_assistant_nemotron_nano_v2_9b.yaml:204]: yes, in the tutorial we will always configure it https://docs.nvidia.com/nemo/gym/latest/tutorials/nemo-rl-grpo/single-node-training.html

i need to unset it generally since not every person needs/wants tool calls or a tool parser plugin

Comment (bxyu-nvidia) [reply examples/nemo_gym/grpo_workplace_assistant_nemotron_nano_v2_9b.yaml:204]: do you have a suggestion on a better comment?

File: nemo_rl/utils/logger.py (hunk 2 above)
Comment (bxyu-nvidia) [reply nemo_rl/utils/logger.py:209]: i am not 100% sure. and those logs also ended up being a lot less useful in my debugging, so i'm even down to remove the logging there.

File: nemo_rl/utils/logger.py (hunk 3 above)
Comment (bxyu-nvidia) [reply nemo_rl/utils/logger.py:975]: https://github.com/NVIDIA-NeMo/RL/blob/f0b1a91dbd67ab55e78d130ef594c8c03637b358/examples/nemo_gym/run_grpo_nemo_gym.py#L152

i forgot to upstream it previously since we weren't officially supporting just trajectory collection previously

File: nemo_rl/algorithms/grpo.py [hunk 4]
```diff
@@ -1267,12 +1325,30 @@ def grpo_train(
                         continue
                     advantages = (rewards - baseline).unsqueeze(-1)
 
+                    baseline: torch.Tensor
+                    metrics["baseline_reward/histogram"] = Histogram(baseline.numpy())
```
Comment (yfw) [inline nemo_rl/algorithms/grpo.py]: I think we should use the logger's `log_histogram` for this instead so other (non-wandb) loggers would be handled correctly.

File: examples/nemo_gym/grpo_workplace_assistant_nemotron_nano_v2_9b.yaml (hunk 1 above)
Comment (yfw) [reply examples/nemo_gym/grpo_workplace_assistant_nemotron_nano_v2_9b.yaml:204]: Ah does it not work to set it to "" by default when we don't need the tool parser plugin?

Comment (bxyu-nvidia) [reply examples/nemo_gym/grpo_workplace_assistant_nemotron_nano_v2_9b.yaml:204]: no it doesn't unfortunately. vllm will error on import 💀

File: nemo_rl/algorithms/grpo.py (hunk 4 above)
Comment (bxyu-nvidia) [reply nemo_rl/algorithms/grpo.py]: Fixed, and also added in advantage logging too at the request of JK.
image

W&B: https://wandb.ai/nvidia/bxyu-nemo-gym-rl-integration/runs/c4460reh?nw=nwuserbxyu
